kernel/src/process/program_loader/elf/load_elf.rs

```diff
@@ -225,66 +217,172 @@ fn map_segment_vmos(
+        let type_ = program_header.get_type().map_err(|_| {
+            Error::with_message(Errno::ENOEXEC, "Failed to parse the program header")
+        })?;
+        if type_ == program::Type::Load {
+            check_segment_align(program_header)?;
+
+            // Update data segment range, not including .bss part.
+            if program_header.flags.is_write() {
+                if program_header.virtual_addr < start_data {
+                    start_data = program_header.virtual_addr;
+                }
+                if end_data < program_header.virtual_addr + program_header.file_size {
+                    end_data = program_header.virtual_addr + program_header.file_size;
+                }
+            }
 
-        if elf_va_range.start < VMAR_LOWEST_ADDR {
-            return_errno_with_message!(Errno::EPERM, "the mapping address is too small");
+            let map_at = relocated_range
+                .relocated_addr_of(program_header.virtual_addr as Vaddr)
+                .expect("Address not covered by `get_range_for_all_segments`");
+
+            map_segment_vmo(program_header, elf_inode, vmar, map_at)?;
         }
+    }
 
-        elf_va_range.clone()
-    };
+    Ok(ElfMappedInfo {
+        full_range: relocated_range,
+        data_segment_size: end_data.saturating_sub(start_data) as usize,
+    })
+}
 
-    let relocated_range = RelocatedRange::new(elf_va_range, map_range.start)
-        .expect("`map_range` should not overflow");
+/// A virtual range and its relocated address.
+pub struct RelocatedRange {
+    original_range: Range<Vaddr>,
+    relocated_start: Vaddr,
+}
 
-    let Some(elf_vmo) = elf_inode.page_cache() else {
-        return_errno_with_message!(Errno::ENOEXEC, "the executable has no page cache");
-    };
-    for loadable_phdr in elf.loadable_phdrs() {
-        let map_at = relocated_range
-            .relocated_addr_of(loadable_phdr.virt_range().start)
-            .expect("`calc_total_vaddr_bounds()` should cover all segments");
-        map_segment_vmo(loadable_phdr, &elf_vmo, vmar, map_at)?;
+impl RelocatedRange {
+    /// Creates a new `RelocatedRange`.
+    ///
+    /// If the relocated address overflows, it will return `None`.
+    pub fn new(original_range: Range<Vaddr>, relocated_start: Vaddr) -> Option<Self> {
+        relocated_start.checked_add(original_range.len())?;
+        Some(Self {
+            original_range,
+            relocated_start,
+        })
+    }
+
+    /// Gets the relocated address of an address in the original range.
+    ///
+    /// If the provided address is not in the original range, it will return `None`.
+    pub fn relocated_addr_of(&self, addr: Vaddr) -> Option<Vaddr> {
+        if self.original_range.contains(&addr) {
+            Some(addr - self.original_range.start + self.relocated_start)
+        } else {
+            None
+        }
     }
+}
 
-    Ok(relocated_range)
+/// The information of mapped ELF segments.
+struct ElfMappedInfo {
```

Comment:
In the latest version, this type should be defined before `map_segment_vmos`. You should be able to understand it after checking the latest code in `load_elf.rs`.